migrate/migrate_users.py

```diff
@@ -188,20 +212,62 @@ def extract_user_names(user_list_file) -> List[str]:
     return empty_lines_removed
 
 
-def extract_merged_team_memberships(org_export_files: list) -> Dict[str, Set[str]]:
+@dataclass(frozen=True)
+class Team:
+    """
+    A class to represent the info weneed to know about each
+    github team to be able to add a user to the team.
+    """
+
+    team_id: int
+    slug: str
+    role: str
+
+
+def extract_merged_team_memberships(
+    org_export_files: list,
+    team_slugs_to_ids: Dict[str, int],
+) -> Dict[str, Set[Team]]:
     """
     Given a list of handles to org export files, return a merged mapping of
-    team slugs to sets of usernames.
+    usernames to all teams they are members of.
     """
+    # Build a mapping from usernames to team slugs.
+    users_to_teams: Dict[str, Set[Team]] = defaultdict(set)
 
-    team_users: Dict[str, Set[str]] = defaultdict(set)
     for json_file in org_export_files:
         org_export_data = json.load(json_file)
         for team in org_export_data["teams"]:
+            team_slug = team["slug"]
+
+            member_team = Team(
+                team_id=team_slugs_to_ids[team_slug],
+                slug=team_slug,
+                role="member",
+            )
+            maintainer_team = Team(
+                team_id=team_slugs_to_ids[team_slug],
+                slug=team_slug,
+                role="maintainer",
+            )
+
             # If the team is in multiple export files, we combine
             # them into one team for the destination org.
-            team_users[team["slug"]].update(team["members"])
-    return team_users
+            for member in team["members"]:
+                if maintainer_team in users_to_teams[member]:
+                    continue
+                    # because the user is already on the team with maintainer level access
+                    # via another org.
+                users_to_teams[member].add(member_team)
+
+            for maintainer in team.get("maintainers", []):
+                if member_team in users_to_teams[maintainer]:
+                    users_to_teams[maintainer].remove(member_team)
+                    # because we're gonna elevate this user's access
+
+                users_to_teams[maintainer].add(maintainer_team)
```

Comment:
`members` and `maintainers` in the export are disjoint, so you could simplify this logic if you wanted to.